src/iceberg/literal.h

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include <compare>
+#include <memory>
+#include <string>
+#include <variant>
+#include <vector>
+
+#include "iceberg/result.h"
+#include "iceberg/type.h"
+
+namespace iceberg {
+
+/// \brief PrimitiveLiteral is a literal value that is associated with a primitive type.
+class ICEBERG_EXPORT PrimitiveLiteral {
```

Comment:
Should we simply use `Literal` instead of `PrimitiveLiteral` which looks lengthy?

Reply:
I afraid in the future there would be MapLiteral and StructLiteral. But all is ok to me

Reply:
MapLiteral and StructLiteral may be combination of Literals and the majority use case is still for primitive types. So I believe making them shorter looks better. @Fokko WDYT?

Reply:
I think `Literal` is more appropriate. Also, there is always an option to introduce a `PrimitiveLiteral` in the class hierarchy. PyIceberg and Iceberg-Java don't have complex types of literals, so I think we're safe there (until it happens).